### PR TITLE
url: accept "any length" credentials for proxy auth

### LINKED ISSUE #5448: cURL does not pass proxy passwords greater than 255 characters to http proxies
### I did this

export http_proxy=http://will:01234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789@proxyserver:12000
curl -v sydonis.com > /dev/null

### I expected the following

I expected cURL to send the entire username/password string to the proxy in the Proxy-Authorization header. Instead cURL appears to send just the username and a colon:

Proxy-Authorization: Basic d2lsbDo=

### curl/libcurl version

curl 7.64.0

### operating system

Debian 10, using a newer kernel: 5.5.0-0.bpo.2 (5.5.17-1~bpo10+1)

### PR DESCRIPTION
**Pull request by bagder**
They're only limited by the maximum string input restrictions, not to
256 bytes.

Added test 1178 to verify

Reported-by: Will Roberts
Fixes #5448

### CODE REVIEW

File: lib/url.c [hunk 1]
```diff
@@ -2355,21 +2355,10 @@ static CURLcode parse_proxy(struct Curl_easy *data,
 static CURLcode parse_proxy_auth(struct Curl_easy *data,
                                  struct connectdata *conn)
 {
-  char proxyuser[MAX_CURL_USER_LENGTH]="";
-  char proxypasswd[MAX_CURL_PASSWORD_LENGTH]="";
+  char *proxyuser = data->set.str[STRING_PROXYUSERNAME];
+  char *proxypasswd = data->set.str[STRING_PROXYPASSWORD];
```
Comment (Lekensteyn) [inline lib/url.c:2359]: Are there any guarantees that both are non-NULL? If not, that could make `Curl_urldecode` crash.

Comment (bagder) [reply lib/url.c:2359]: I figured the tests would crash and burn then... =)

Comment (bagder) [reply lib/url.c:2359]: The function `parse_proxy_auth()`  is only called if `conn->bits.proxy_user_passwd` is set, which presumably never happens if they are NULL.

Comment (bagder) [reply lib/url.c:2359]: Hm, the password can actually still be NULL...

Comment (bagder) [reply lib/url.c:2359]: Adding checks

File: lib/urldata.h [hunk 2]
```diff
@@ -1228,17 +1228,6 @@ typedef enum {
...
 } Curl_RtspReq;
 
-/*
- * Values that are generated, temporary or calculated internally for a
- * "session handle" must be defined within the 'struct UrlState'.  This struct
- * will be used within the Curl_easy struct. When the 'Curl_easy'
- * struct is cloned, this data MUST NOT be copied.
- *
- * Remember that any "state" information goes globally for the curl handle.
- * Session-data MUST be put in the connectdata struct and here.  */
```
Comment (bagder) [inline lib/urldata.h:1238]: This comment block really serves no purpose so I removed it while here. On purpose. Even though not strictly related to the defines below that I wanted to remove.

File: lib/url.c [hunk 3]
```diff
@@ -2355,24 +2355,14 @@ static CURLcode parse_proxy(struct Curl_easy *data,
...
-    strncpy(proxypasswd, data->set.str[STRING_PROXYPASSWORD],
-            MAX_CURL_PASSWORD_LENGTH);
-    proxypasswd[MAX_CURL_PASSWORD_LENGTH-1] = '\0'; /*To be on safe side*/
-  }
-
-  result = Curl_urldecode(data, proxyuser, 0, &conn->http_proxy.user, NULL,
-                          FALSE);
-  if(!result)
+  if(proxyuser)
+    result = Curl_urldecode(data, proxyuser, 0, &conn->http_proxy.user, NULL,
```
Comment (Lekensteyn) [inline lib/url.c:2363]: Shouldn't it clear `conn->http_proxy.user` and `.passwd` if no username/password is available?

Comment (bagder) [reply lib/url.c:2363]: I don't think that should be necessary. We should treat a NULL user there as non-existing in the code. I'll take another check and see that this is true...

**Pull request closed without merging**